### PR TITLE
[debugger enhance] don't block gdbserver thread while executing

### PR DESCRIPTION
**Pull request by xujuntwt95329**

### CODE REVIEW

File: core/iwasm/libraries/debug-engine/debug_engine.c [hunk 1]
```diff
@@ -104,6 +112,43 @@ control_thread_routine(void *arg)
...
+                        debug_inst->stopped_thread->current_status->signal_flag;
+                tid = debug_inst->stopped_thread->handle;
+
+                if (debug_inst->stopped_thread->current_status->running_status
+                    == STATUS_EXIT) {
+                    /* If the thread exits, report "W00" if it's the last thread
+                     * in the cluster, otherwise ignore this event */
+                    status = 0;
+
+                    /* By design, all the other threads should stopped at this
```
Comment (wenyongh) [inline core/iwasm/libraries/debug-engine/debug_engine.c:132]: should stop or should have been stopped?

File: core/iwasm/libraries/debug-engine/debug_engine.c [hunk 2]
```diff
@@ -960,8 +995,15 @@ wasm_debug_instance_kill(WASMDebugInstance *instance)
 
     while (exec_env) {
         wasm_cluster_thread_send_signal(exec_env, WAMR_SIG_TERM);
+        if (instance->current_state == APP_STOPPED) {
+            /* Resume all threads so they can receive the TERM signal */
+            exec_env->current_status->running_status = STATUS_RUNNING;
+            os_cond_signal(&exec_env->wait_cond);
```
Comment (wenyongh) [inline core/iwasm/libraries/debug-engine/debug_engine.c:1001]: Should we wrap os_cond_signal with `os_mutex_lock(&exec_env->wait_lock)` and `os_mutex_unlock(&exec_env->wait_lock)`

File: core/iwasm/libraries/debug-engine/gdbserver.c [hunk 3]
```diff
@@ -125,51 +138,157 @@ handler_packet(WASMGDBServer *server, char request, char *payload)
...
 
+static void
+process_packet(WASMGDBServer *server)
+{
+    uint8 *inbuf = (uint8 *)server->receive_ctx->receive_buffer;
+    char request;
+    char *payload = NULL;
+
+    request = inbuf[0];
+    payload = (char *)&inbuf[1];
```
Comment (wenyongh) [inline core/iwasm/libraries/debug-engine/gdbserver.c:149]: Will the inbuf's length be no smaller than 2? Shall we check the length here?

File: core/iwasm/libraries/debug-engine/gdbserver.c [hunk 4]
```diff
@@ -125,51 +138,157 @@ handler_packet(WASMGDBServer *server, char request, char *payload)
...
+
+    LOG_VERBOSE("receive request:%c %s\n", request, payload);
+    handler_packet(server, request, payload);
```
Comment (wenyongh) [inline core/iwasm/libraries/debug-engine/gdbserver.c:152]: How about changing `handle_packet`?

File: core/iwasm/libraries/debug-engine/gdbserver.c [hunk 5]
```diff
@@ -125,51 +138,157 @@ handler_packet(WASMGDBServer *server, char request, char *payload)
...
+}
+
+static inline void
+push_byte(rsp_recv_context_t *ctx, unsigned char ch, bool checksum)
+{
+    if (ctx->receive_index >= sizeof(ctx->receive_buffer)) {
+        LOG_ERROR("RSP message buffer overflow");
+        bh_assert(false);
```
Comment (wenyongh) [inline core/iwasm/libraries/debug-engine/gdbserver.c:160]: Had better add `return`?

File: core/iwasm/libraries/debug-engine/handler.c [hunk 6]
```diff
@@ -441,14 +438,31 @@ handle_v_packet(WASMGDBServer *server, char *payload)
...
+    /* According to
+       https://sourceware.org/gdb/onlinedocs/gdb/Packets.html#Packets, the "?"
+       package should be sent when connection is first established to query the
+       reason the target halted */
+    bh_assert(debug_inst->current_state == DBG_LAUNCHING);
 
-    tid = wasm_debug_instance_wait_thread(
-        (WASMDebugInstance *)server->thread->debug_instance, tid, &status);
+    /* Waiting for the stop event */
+    while (!debug_inst->stopped_thread) {
```
Comment (wenyongh) [inline core/iwasm/libraries/debug-engine/handler.c:454]: Dead loop occupies a lot CPU resource, is it possible to wait a signal? Or add a timeout wait inside the loop?

File: core/iwasm/libraries/thread-mgr/thread_manager.c [hunk 7]
```diff
@@ -284,6 +284,20 @@ wasm_cluster_del_exec_env(WASMCluster *cluster, WASMExecEnv *exec_env)
...
     bh_assert(exec_env->cluster == cluster);
+
+#if WASM_ENABLE_DEBUG_INTERP != 0
+    /* Wait for debugger control thread to process the
+       stop event of this thread */
+    if (cluster->debug_inst) {
+        /* lock the debug_inst->wait_lock so
+           other threads can't fire stop events */
+        os_mutex_lock(&cluster->debug_inst->wait_lock);
+        while (cluster->debug_inst->stopped_thread == exec_env) {
```
Comment (wenyongh) [inline core/iwasm/libraries/thread-mgr/thread_manager.c:295]: Same as above, can we add os_cond_timedwait(&cluster->debug_inst->wait_cond, ..) inside loop?

File: core/iwasm/libraries/debug-engine/debug_engine.c (hunk 2 above)
Comment (xujuntwt95329) [reply core/iwasm/libraries/debug-engine/debug_engine.c]: Done

File: core/iwasm/libraries/debug-engine/gdbserver.c (hunk 3 above)
Comment (xujuntwt95329) [reply core/iwasm/libraries/debug-engine/gdbserver.c:149]: Add a check for request. If it's '\0', that means this is an empty request, and we ignore it

File: core/iwasm/libraries/debug-engine/gdbserver.c (hunk 4 above)
Comment (xujuntwt95329) [reply core/iwasm/libraries/debug-engine/gdbserver.c]: Done

File: core/iwasm/libraries/debug-engine/gdbserver.c (hunk 5 above)
Comment (xujuntwt95329) [reply core/iwasm/libraries/debug-engine/gdbserver.c:165]: Done

File: core/iwasm/libraries/debug-engine/handler.c (hunk 6 above)
Comment (xujuntwt95329) [reply core/iwasm/libraries/debug-engine/handler.c:455]: Done

File: core/iwasm/libraries/thread-mgr/thread_manager.c (hunk 7 above)
Comment (xujuntwt95329) [reply core/iwasm/libraries/thread-mgr/thread_manager.c:295]: Done, thanks!

File: core/iwasm/libraries/debug-engine/gdbserver.c [hunk 8]
```diff
@@ -119,57 +132,169 @@ wasm_close_gdbserver(WASMGDBServer *server)
...
+    char request;
+    char *payload = NULL;
+
+    request = inbuf[0];
+    payload = (char *)&inbuf[1];
+
+    if (request == '\0') {
+        LOG_VERBOSE("ignore empty request");
+        return;
+    }
```
Comment (wenyongh) [inline core/iwasm/libraries/debug-engine/gdbserver.c:154]: Should move these lines before L149? If inbuf[0] is '\0', we should not read inbuf[1] in L149?

**Pull request merged**